src/identity-providers.js

```diff
@@ -0,0 +1,105 @@
+'use strict'
+const Identity = require('./identity')
+const OrbitDBIdentityProvider = require('./orbit-db-identity-provider')
+const Keystore = require('orbit-db-keystore')
+
+let supportedTypes = {
+  'orbitdb': OrbitDBIdentityProvider,
+}
+
+const getHandlerFor = (type) => {
+  if (!IdentityProviders.isSupported(type)) {
+    throw new Error(`IdentityProvider type '${type}' is not supported`)
+  }
+  return supportedTypes[type]
+}
+
+class IdentityProviders {
+  constructor(options = {}) {
+    this._keystore = options.keystore || Keystore.create(options.keypath || './orbitdb/ipkeys')
+  }
+
+  async create(id, options = {}) {
+    const { publicKey, idSignature, selfSignedPubKeyIdSig } = await this.signIdentity(id, options)
+    const IdentityProvider = getHandlerFor(options.type)
+    const pubKeyIdSignature = await IdentityProvider.signIdentity(publicKey + idSignature, options) || selfSignedPubKeyIdSig
+    return new Identity(id, publicKey, idSignature, pubKeyIdSignature, IdentityProvider.type || type, this)
+  }
+
+  async signIdentity(id, options = {}) {
+    const keystore = options.keystore || this._keystore
+    const key = await keystore.getKey(id) || await keystore.createKey(id)
+    const publicKey = await key.getPublic('hex')
+    const idSignature = await keystore.sign(key, id)
+    const selfSignedPubKeyIdSig = await keystore.sign(key, publicKey + idSignature)
+    return { publicKey, idSignature, selfSignedPubKeyIdSig }
+  }
+
+  async sign (identity, data) {
+    const signingKey = await this._keystore.getKey(identity.id)
+    if (!signingKey)
+      throw new Error(`Private signing key not found from Keystore`)
+
+    const signature = await this._keystore.sign(signingKey, data)
+    return signature
+  }
+
+  async verify (signature, publicKey, data) {
+    return this._keystore.verify(signature, publicKey, data)
+  }
+
+  async verifyIdentity (identity, options = {}) {
+    const verified = await this._keystore.verify(
+      identity.signatures.id,
+      identity.publicKey,
+      identity.id
+    )
+
+    return verified && await IdentityProviders.verifyIdentity(identity, options)
+  }
+
+  static isSupported (type) {
+    return Object.keys(supportedTypes).includes(type)
+  }
+
+  static addIdentityProvider (options) {
+    if (!options.IdentityProvider) {
+      throw new Error('IdentityProvider class needs to be given as an option')
+    }
+
+    if (!options.IdentityProvider.type ||
+      typeof options.IdentityProvider.type !== 'string') {
+      throw new Error('Given IdentityProvider class needs to implement: static get type() { /* return a string */}.')
+    }
+
+    supportedTypes[options.IdentityProvider.type] = options.IdentityProvider
+  }
+
+  static addIdentityProviders (options) {
```

Comment:
I'm unsure if this method is useful to add. It does the same as if user would call:
```javascript
    identityProviders.forEach((identityProvider) => {
      IdentityProviders.addIdentityProvider({ IdentityProvider: identityProvider })
    })
```

Note also that (I believe) the argument here is incorrectly `addIdentityProvider` as it should be `IdentityProvider`. Also, as the function processes and *array* of IdentityProvider objects, it's be good if the function argument was an array instead of an object that contains an array.

I would remove it and let user handle any looping on the app level and just provide `addIdentityProvider` to add a single IP at a time.